### ISSUE TITLE
How do I need to do to work early the sensor?

### ISSUE THREAD

**Issue by harinezumi367** [ev3dev/ev3dev#231]
Connect the NXT light sensor, where it measured the update rate values, it has not been only updated at a rate of about 80Hz from 60.
How do I need to do to work more quickly?

**Comment by laurensvalk** [comment]
I am working on a program for a self balancing robot using a gyro sensor. I found that the update rate of the analog value (HiTechnic Gyro) is about 100 Hz so I started searching and found this thread. I thought I'd ask my question here as it is essentially the same as the original question.

The final robot should probably be able to balance with a 100 Hz loop, but for initial modelling of the robot and controller design, it would be really useful to have all the information we can get and poll at 1ms  / 1000 Hz.

@dlech, you mentioned you might be planning to make this value adjustable. I wonder if you have had a chance to do this yet? If so, where can I find the file to change this update rate? I haven't attempted to recompile the kernel with the adjusted update rate yet, but maybe I don't need to :)

![gyropolling-analog](https://cloud.githubusercontent.com/assets/12326241/7533213/d7c367d4-f56f-11e4-863c-589e1ca9252d.png)

![gyropolling-legoev3](https://cloud.githubusercontent.com/assets/12326241/7533219/ed6c0898-f56f-11e4-8446-4e3f4bd62f93.png)

The data attached here was obtained by simply reading the relevant sensor value file approximately once per 1.6 ms, without any particular API.  (Loop with 0.6 ms to read/write sensors and motors, and 1ms sleep). The red stars are the samples. The blue line is just a first order hold. We'll probably be using the LEGO gyro for now.

By the way, great work on ev3dev! Everything I tried so far simply just works :)

**Comment by dlech** [comment]
> I wonder if you have had a chance to do this yet?

Not yet...

> If so, where can I find the file to change this update rate?

Looks like the link above is broken. The code now lives [here](https://github.com/ev3dev/lego-linux-drivers/blob/master/ev3/legoev3_analog.c#L582). You just need to change `UPDATE_SLOW_NS` to `UPDATE_FAST_NS`.

> The data attached...

From the looks of the HiTechnic graph, it looks like you are using the generic analog driver. There is a driver for the HiTechnic gyro sensor that scales the value to deg/sec.

> By the way, great work on ev3dev! Everything I tried so far simply just works :)

Awesome!

**Comment by laurensvalk** [comment]
Thanks for your response. I'll do some experiments with the various gyro sensors in the coming weeks before assuming a specific scaling.

By the way, at a quick glance, the GYRO-FAS mode in the LEGO Gyro is a lower resolution mode that has a wider measurement range. While making similar movements as in the graphs posted above (which is clearly capped at +/-440), I didn't reach the cap on the GYRO-FAS mode.

I'll report back in a different thread with proper measurements and scaling between the modes later so it can be added to the documentation/wiki.

**Comment by laurensvalk** [comment]
I recompiled the kernel with the change suggested in this thread. I believe it's running:

segway@ev3dev:~$ uname -a 
Linux ev3dev 3.16.7-ckt10-5-rc1-ev3dev-ev3+ #1 PREEMPT Wed May 13 10:19:01 CEST 2015 armv5tejl GNU/Linux

However, the sensor value still stays constant for 10ms. Do you know of any further changes I might have to make or checks I can do to verify that the change has been implemented correctly?

**Comment by laurensvalk** [comment]
Well, my previous comment wasn't completely accurate. When the analog device was eventually polling at the fast rate, I had been changing a few suggested parameters at the same time. 

Before closing this thread, I went back to compile the kernel systematically after each change to see which change was responsible, so you can possibly find the cause more easily.

Change 1) 
At first, I changed  UPDATE_SLOW_NS to UPDATE_FAST_NS as suggested earlier in this thread. This didn't help. 

Change 2) Instead, I tried changing this section in legoev3_analog.c

#define UPDATE_SLOW_NS		10000000		/*  10 msec */
#define UPDATE_FAST_NS		(UPDATE_SLOW_NS / 10)	/*   1 msec */

to this

#define UPDATE_SLOW_NS		1000000		/*  1 msec */
#define UPDATE_FAST_NS		(UPDATE_SLOW_NS )	/*   1 msec */

Which did work.

Change 3) I also set HZ to 1000. Surprisingly, it wasn't necessary to set HZ to 1000 to get the ADC to be polled more often. (Though setting Hz to 1000 is probably still useful for more accurate timing.)

In summary, I tried the following configurations, with these results:

Changes: none --> ADC update rate >= 10 ms
Changes: 1 --> ADC update rate >= 10 ms
Changes: 1+3 --> ADC update rate >= 10 ms
Changes: 2 --> ADC update rate < 10 ms
Changes: 2+3 --> ADC update rate < 10 ms

This is all with the May 1 image + kernel/driver code from the May 1 release. I'm not sure why change 3 works while change 2 does not, but I hope this helps.

**Comment by dlech** [comment]
Looks like I pointed you to the wrong line in the code, `UPDATE_SLOW_NS` is also used [here](https://github.com/ev3dev/lego-linux-drivers/blob/master/ev3/legoev3_analog.c#L227). Changing the `#define` of course changes the value everywhere, so no mystery here.

**Comment by laurensvalk** [comment]
The original issue by  @harinezumi367 and the follow up issue about the analog polling frequency is technically solved, so I think we can close this issue.

Before we close it, I'll add a feature request/enhancement. That is, it would be ideal if the update rate can be changed through something similar to the poll_ms attribute, which is currently unused for analog sensors.

The existing poll_ms attribute can even be used if the new lower bound is 1ms, which is probably sufficient. The existing 10ms can be the default.

**Comment by dlech** [comment]
Sounds good. I'll wait for you to open the new issue to share my thoughts on the subject.
